mozetl/taar/taar_amodump.py

```diff
@@ -0,0 +1,370 @@
+                # Skip this record
+                logger.error(e)
+        return page_urls
+
+
+class Undefined:
+    """
+    This value is used to disambiguate None vs a non-existant value on
+    dict.get() lookups
+    """
+    pass
+
+
+def marshal(value, name, type_def):
+    serializers = {typing.List: list,
+                   typing.Dict: dict,
+                   str: unicode,
+                   unicode: unicode,
+                   int: int,
+                   float: float,
+                   bool: bool}
+
+    if issubclass(type_def, JSONSchema):
+        obj = {}
+        for attr_name, attr_type_def in type_def.meta.items():
+            attr_value = value.get(attr_name, Undefined)
+            if attr_value is not Undefined:
+                # Try marshalling the value
+                obj[attr_name] = marshal(attr_value,
+                                         attr_name,
+                                         attr_type_def)
+        return obj
+    elif issubclass(type_def, typing.Container) and type_def not in [str, bytes]:
+        if issubclass(type_def, typing.List):
+            item_type = type_def.__args__[0]
+            return [marshal(j, name, item_type) for j in value]
+        elif issubclass(type_def, typing.Dict):
+            if value is None:
+                return None
+            k_cast, v_cast = type_def.__args__
+            dict_vals = [(marshal(k, name, k_cast),
+                          marshal(v, name, v_cast))
+                         for k, v in value.items()]
+            return dict(dict_vals)
+    else:
+        return serializers[type_def](value)
+
+
+@click.command()
+@click.option("--path", default=tempfile.mkdtemp())
+@click.option("--date", required=True)
+@click.option("--workers", default=100)
+def main(path, date, workers):
+    amodb = AMODatabase(path, int(workers))
+    amodb.fetch_pages()
+    amodb.fetch_versions()
+    try:
+        large_blob = {}
+
+        successful = 0
+        for i, fname in enumerate(os.listdir(amodb._amo_cache_dir)):
+            guid = fname.split(".json")[0]
+            client_blob = amodb.parse_file(guid)
+
+            if client_blob is None:
+                continue
+
+            large_blob[guid] = client_blob
+            successful += 1
+
+            if i > 0 and i % 200 == 0:
+                sys.stdout.write('.')
+                sys.stdout.flush()
+
+        fail_rate = 1.0 - (successful * 1.0 / len(os.listdir(amodb._amo_cache_dir)))
+        if fail_rate >= 0.01:
+            logger.error("Unexpectedly high failure rate for merging versions with addon JSON.  %0.4f" % (fail_rate))
+            sys.exit(1)
+        else:
+            logger.info("Version merging failure rate: %0.4f" % fail_rate)
+
+        store_json_to_s3(json.dumps(large_blob),
+                         AMO_DUMP_FILENAME,
+                         date,
+                         AMO_DUMP_PREFIX,
+                         AMO_DUMP_BUCKET)
```

Comment:
It might be useful to make bucket and prefix configurable, for testing without writing to the production location,